syntax = "proto3";
import "Common/Anime.proto";

// ===========================> ANIMELIST PROTO FILE <===========================
// This file contains the proto file for the AnimeList service.
// The AnimeList service is responsible for asking the AnimeRepository for anime data,
// Transforming it how ever required and sending it as a response for the Controller to use.
// It can also implement some logic over the data before sending it back.
// ==============================================================================

// Used when user wants to list all animes
message get_all_animes {} // --------------------------> REQUEST FOR GETTING ALL ANIMES

message get_all_animes_Response {
    repeated Anime animes = 1;
}

// Used when user wants to get all the information on a specific anime
message get_anime_by_name_Request { // ----------------> REQUEST FOR GETTING ANIME BY NAME
    string anime_name = 1;
}

message get_anime_by_name_Response {
    Anime anime = 1;
}

// Used when user wants to get all the information on multiple animes that he specifies
message get_multiple_anime_by_name_Request { // -------------> REQUEST FOR GETTING MORE THAN ONE ANIME
    repeated string anime_names = 1; // List of anime names to get
}

message get_multiple_anime_by_name_Response {
    repeated Anime animes = 1;
}

// Used when user wants to get recommendations of similar animes to the one he specifies
message get_similar_anime_Request { // -------------> REQUEST FOR GETTING A LIST OF ANIMES SIMILAR TO THE ONE PROVIDED
    string anime_name = 1;
}

message get_similar_anime_Response {
    repeated Anime animes = 1;
}

// Used when user wants to get a list of animes that he may like
message recomended_animeList_Request { // -------------> REQUEST FOR GETTING RECOMENDED ANIMES FOR A USER
    repeated Anime animes_most_liked = 1;
}

message recomended_animeList_Response {
    repeated Anime animes = 1;
}

service AnimeList { // --------------------------> SERVICE FOR ANIMELIST
    rpc GetAllAnimes (get_all_animes) returns (get_all_animes_Response);
    rpc GetAnimeByName (get_anime_by_name_Request) returns (get_anime_by_name_Response);
    rpc GetMultipleAnimeByName (get_multiple_anime_by_name_Request) returns (get_multiple_anime_by_name_Response);
    rpc GetSimilarAnime (get_similar_anime_Request) returns (get_similar_anime_Response);
    rpc GetRecomendedAnimeList (recomended_animeList_Request) returns (recomended_animeList_Response);
}